Add initial proposal of introducing Knative Eventing Mesh (#386)
kyma-project/community: sig-and-wg/sig-core/proposals/introducing_knative_brokers.md

# Integration of the Knative-based function controller

## Introduction

TBD

## Features

### Current Features

- As a developer, I want to publish events to Kyma from outside the cluster.
- As a developer, I want to publish events to Kyma from inside the cluster.
- As a developer, I want to subscribe to events in Kyma from a specific source with a specific event-type.
- As an admin, I want to enable or disable events in a specific namespace.

> ### New features

Reviewer: This is not a new feature. Right now, the cm `default-ch-webhook` exists which enables this functionality.

> # Integration of the Knative-based function controller

Reviewer 2: The title should be about introducing Knative brokers in Kyma.

### Solution 1: Knative Brokers Only

#### Overview

The usage of Knative Eventing shifts the eventing paradigm in Kyma, from a single messaging Pub/Sub system, to an eventing mesh with isolated fault domains, access control and dynamic event routing where many senders inject events into the mesh from multiple source points, and many sinks (i.e subscribers) receive all or a subset of those events based on filters and access control. The responsiblity of Kyma system is to correctly and dynamically wire the eventing mesh to match the user eventing needs and achieve optimum event routing.

Solution 1 describes how Kyma system will build and wire the eventing mesh to achieve proper event routing from external application to the subscribers such as the lambda functions.

#### Architecture

For every namespace that is allowed to receive events, there is a default `Broker` in the namespace. Kyma will do all the wiring from the `Event Service` to the local namespace `Brokers` and subscribers who want to receive events, should create a Knative `Trigger` against their local default `Broker`.

![architecture](assets/knative_brokers_arch.svg)

> 1. For every new `Application` the `Application Operator` creates an `Event Service` and its corresponding  `Knative Channel` that will be used to fanout events to all allowed brokers.

Reviewer: >  if you use the SubscriberSpec.Ref so you would have to use the URI instead 
Exactly, we are planning to use the URI here.
```
apiVersion: eventing.knative.dev/v1alpha1
kind: Subscription
metadata:
  name: broker-sub-test
  namespace: kyma-integration
spec:
  channel:
    apiVersion: messaging.knative.dev/v1alpha1
    kind: Channel
    name: test-app-ch
  reply: {}
  subscriber:
    uri: http://default-broker.<target-namespace>/
```

2. For every new namespace, Kyma Console ensures it has the label `knative-eventing-injection=enabled` so that a default `Broker` gets created by the Knative Eventing controller.
3. Once the `Application Broker` gets a provision request for a `Service Instance` which has `bindable` parameter set to `false`, it will create a `Knative Subscription` that has:
   - The channel pointing to the relevant `Event Service` fanout `Channel`
   - The subscriber pointing to the default `Broker` of the namespace
   - The ownerRef set as the `Service Instance` that just got provisioned which ensures the subscription gets deleted by Kubernetes garbage collector when the `Service Instance` gets deprovisioned.
4. `Kyma Console` creates the proper `Trigger` objects for the `Lambda` function for every event type chosen as a lambda trigger.

> 5. Events received by `Event Service` will be delivered to its channel which in turn will do a fanout to all connected brokers which will finally deliverd it to the lambda functions.

Reviewer 3: nit 'delivered' (the second one is deliverd)

# Integration of the Knative-based function controller

## Introduction

> As of 0.5, Knative Eventing defines [`Broker` and `Trigger`](https://knative.dev/docs/eventing/broker-trigger/) objects as higher level abstractions that allows brokering events based on some event filtering criterea.

Reviewer: ```suggestion
As of 0.5, Knative Eventing defines [`Broker` and `Trigger`](https://knative.dev/docs/eventing/broker-trigger/) objects as higher-level abstractions that allow brokering events based on some event filtering criteria.
```

> This is a proposal of how to introduce these objects in Kyma as an alternative to using the existing Kyma Event Bus. It describes how to build namespace scoped eventing mesh with fault domain isotation and access control. The proposal describes two possible solutions, the migration path and tries to highlight and address the challenges and open questions of using them.

Reviewer: ```suggestion
This is a proposal of how to introduce these objects in Kyma as an alternative to using the existing Kyma Event Bus. It describes how to build namespace scoped eventing mesh with fault domain isolation and access control. The proposal describes two possible solutions, the migration path and tries to highlight and address the challenges and open questions of using them.
```

## Features

### Current Features

- As a developer, I want to publish events to Kyma from outside the cluster.
- As a developer, I want to publish events to Kyma from inside the cluster.
- As a developer, I want to subscribe to events in Kyma from a specific source with a specific event-type.
- As an admin, I want to enable or disable events in a specific namespace.

### New features

- As an admin, I want to configure the default cluster wide Knative `Channel` kind to be used. 
- As an admin, I want to configure a Knative `Channel` kind to be used by default for events per namespace that overrides the cluster default configuration.

## Proposed solution

> There are two proposed solutions. Solution 1 is an instant transition and deprecation proposal where the new `Broker` based eventing will replace the existing Event Bus. The Kyma `Susbscription` and `EventActivation` CRDs will be removed and Event Bus users will need to migrate to the new knative CRDs.

Reviewer: - `new Broker/Trigger` based
- `EventActivation` CRDs will not be removed now until we solve the Event registry workflow.

> Solution 2 is about having both systems coexisting for one or two releases, during which the namespaces with the `knative-injection-enabled=true` label will have the new `Broker` based eventing, while the ones that don't will just have the same old Kyma Event Bus based behavior (i.e uses Kyma `EventActivation` and `Subscription` CRDs) in order to:

Reviewer: ```suggestion
Solution 2 is about having both systems coexisting for one or two releases, during which the namespaces with the `knative-eventing-injection=enabled` label will have the new `Broker` based eventing, while the ones that don't will just have the same old Kyma Event Bus based behavior (i.e uses Kyma `EventActivation` and `Subscription` CRDs) in order to:
```
Let's be consistent with the label.

1. For every new `Application` the `Application Operator` creates an `Event Service` and its corresponding  `Knative Channel` that will be used to fanout events to all allowed brokers.
2. For every new namespace, Kyma Console ensures it has the label `knative-eventing-injection=enabled` so that a default `Broker` gets created by the Knative Eventing controller.
3. Once the `Application Broker` gets a provision request for a `Service Instance` which has `bindable` parameter set to `false`, it will create a `Knative Subscription` that has:
   - The channel pointing to the relevant `Event Service` fanout `Channel`
   - The subscriber pointing to the default `Broker` of the namespace
   - The ownerRef set as the `Service Instance` that just got provisioned which ensures the subscription gets deleted by Kubernetes garbage collector when the `Service Instance` gets deprovisioned.
4. `Kyma Console` creates the proper `Trigger` objects for the `Lambda` function for every event type chosen as a lambda trigger.
5. Events received by `Event Service` will be delivered to its channel which in turn will do a fanout to all connected brokers which will finally deliverd it to the lambda functions.

> 

Reviewer: I think it would make sense to add the deprovisioning process in a different section. We need to answer:
- Who deletes the Knative Subscription to the fanout channel? IMO the `application` should be the owner of the Knative Channel.

In the Knative Brokers approach, it is not clear how are we solving this:
> - As a developer, I want to publish events to Kyma from inside the cluster.

#### Architecture

![architecture](assets/knative_brokers_co_exist_arch.svg)

1. For every new `Application` the `Application Operator` creates an `Event Service` and its corresponding  `Knative Channel` that will be used to fanout events to the opted-in namespaces.
2. From Kyma Console UI, when the user creates a namespace, the pop up will show an info message and toggle with something like:
>Kyma now comes with a new `Knative Broker` based eventing feature (beta).
>Would you like to try it out for this namespace?. Read more about it here <link to more docs>
and by enabling the toggle, Kyma Console ensures it has the label `knative-eventing-injection=enabled` so that a default `Broker` gets created by the Knative Eventing controller.
3. Once the `Application Broker` gets a provision request for a `Service Instance` which has `bindable` parameter set to `false`, it creates a `Knative Subscription` for `opted-in` namespaces (as in solution 1) or an `EventActivation` for `legacy namespaces`.
4. `Kyma Console` creates the appropriate `Trigger` or `Kyma Subscription` objects for the `Lambda` function for `opted-in` and `legacy` namespaces.

> 5. Publishing events from `Publish` will deliver events to subscribers in both `opted-in` and `legacy` namespaces.

Reviewer: I have a few concerns with proposal 2:
- Publish service now needs to be a controller and needs to keep a state of all the new knative channels(for all applications) created to fanout the events. This is something I would avoid.
- What will be the response when one of the publish fails? E.g. there are 2 channels for 2 applications and a channel(old one). Will we fail the publish request when 1 channel fails and rest succeed? What are the implications?
- There should be another migration plan for this approach to a purely new approach which should be here IMO.
- We are not eliminating Publish Service but adding more logic to it which is against our objective.
